diagramming-tool/src/Component/Home/Home.js

```diff
@@ -0,0 +1,274 @@
+        ctx.lineTo(shape.x + shape.width, shape.y + shape.height / 2);
+        ctx.lineTo(shape.x + shape.width / 2, shape.y + shape.height);
+        ctx.lineTo(shape.x, shape.y + shape.height / 2);
+        ctx.closePath();
+        ctx.fillStyle = "white";
+        ctx.fill();
+        ctx.stroke();
+      }
+    });
+  }, [shapes]);
+
+  useEffect(() => {
+    draw();
+  }, [draw]);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+
+    const handleMouseDown = (e) => {
+      const rect = canvas.getBoundingClientRect();
+      const x = e.clientX - rect.left;
+      const y = e.clientY - rect.top;
+
+      const clickedShape = shapes.find((shape) => {
+        if (
+          shape.type === ShapeTypes.RECTANGLE ||
+          shape.type === ShapeTypes.SQUARE
+        ) {
+          return (
+            x >= shape.x &&
+            x <= shape.x + shape.width &&
+            y >= shape.y &&
+            y <= shape.y + shape.height
+          );
+        } else if (shape.type === ShapeTypes.CIRCLE) {
+          return (
+            Math.sqrt((x - shape.x) ** 2 + (y - shape.y) ** 2) <= shape.radius
+          );
+        } else if (shape.type === ShapeTypes.DIAMOND) {
+          const centerX = shape.x + shape.width / 2;
+          const centerY = shape.y + shape.height / 2;
+          const dx = Math.abs(x - centerX);
+          const dy = Math.abs(y - centerY);
+          return dx / (shape.width / 2) + dy / (shape.height / 2) <= 1;
+        }
+        return false;
+      });
+
+      if (clickedShape) {
+        isDrawing.current = true;
+        draggedShape.current = clickedShape;
+        dragOffset.current = { x: x - clickedShape.x, y: y - clickedShape.y };
+        setSelectedShape(clickedShape.id);
+      }
+    };
+
+    const handleMouseMove = (e) => {
+      if (isDrawing.current) {
+        const rect = canvas.getBoundingClientRect();
+        const x = e.clientX - rect.left - dragOffset.current.x;
+        const y = e.clientY - rect.top - dragOffset.current.y;
+        draggedShape.current.x = x;
+        draggedShape.current.y = y;
+        draw();
+      }
+    };
+
+    const handleMouseUp = () => {
+      isDrawing.current = false;
+    };
+
+    canvas.addEventListener("mousedown", handleMouseDown);
+    canvas.addEventListener("mousemove", handleMouseMove);
+    canvas.addEventListener("mouseup", handleMouseUp);
+
+    return () => {
+      canvas.removeEventListener("mousedown", handleMouseDown);
+      canvas.removeEventListener("mousemove", handleMouseMove);
+      canvas.removeEventListener("mouseup", handleMouseUp);
+    };
+  }, [shapes, draw]);
+
+  const addShape = (type) => {
+    const newShape = {
+      id: Date.now(),
+      type,
+      x: 100,
+      y: 100,
+      width: 100,
+      height: 100,
+      radius: 50,
+      size: 80,
+    };
+    setShapes([...shapes, newShape]);
+    setSelectedShape(newShape.id);
+  };
+// 
+  const handleUndo = () => {
+  };
```

Comment:
Complete the logic for undo  save,delete,Redo

Reply:
Done changes